base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,206 @@
+
+        # Note: Integer conversion may cause inaccuracy
+        self.step = int(ramp_interval * dds.sysclk / 4.0)
+
+        self.ramp_mode = ramp_mode
+
+
+class RAMProfileMap:
+    def __init__(self, core):
+        """Initialize a RAM profile to DDS mapping
+
+        Args:
+            core: The ARTIQ Core instance for time control.
+        """
+        self.ram_profile_map = []
+        self.cplds = []
+        self.core = core
+
+    def append(self, dds, ram_profile):
+        """Append a RAM profile into the builder. In addition, register the
+            CPLDs that have DDSes that playback the RAM profile.
+
+        Args:
+            dds: The DDS that will playback the RAM profile.
+            ram_profile: The RAM profile.
+        """
+        self.ram_profile_map.append((dds, ram_profile))
+
+        # Add the associlated CPLD to the list if not already there.
+        # ARTIQ-Python does not support iterration of a Python builtin set
+        # The workaround is to maintain a list, but convert it into set to
+        # avoid CPLD duplications.
+        cplds = set(self.cplds)
+        cplds.add(dds.cpld)
+        self.cplds = list(cplds)
+
+    @kernel
+    def program(self):
+        """Program the RAM profiles to the corresponding AD9910s.
+
+        """
+        # Switch to profile 0 for RAM. This is the default profile.
+        # Using other profiles for RAM is possible with appropriate parameters
+        # for future AD9910 function calls.
+        for cpld in self.cplds:
+            cpld.set_profile(0)
+
+        for dds, ram_profile in self.ram_profile_map:
+            # Datasheets strongly recommands setting ram_enable=0 before
+            # writing anything to the RAM profiles
+            dds.set_cfr1(ram_enable=0)
+            dds.cpld.io_update.pulse_mu(8)
+
+            # Write RAM profile that corresponds to the DDS
+            dds.set_profile_ram(
+                start=ram_profile.start_addr,
+                end=ram_profile.end_addr,
+                step=ram_profile.step,
+                mode=ram_profile.ramp_mode)
+            dds.cpld.io_update.pulse_mu(8)
+
+            # Program the RAM, break_realtime to avoid RTIOunderflow
+            dds.write_ram(ram_profile.ram)
+            self.core.break_realtime()
+
+            # Alternatively, use dds.set()
+            dds.set_mu(ftw=ram_profile.ftw,
+                       pow_=ram_profile.pow,
+                       asf=ram_profile.asf,
+                       ram_destination=ram_profile.dest)
+
+        # Queue in RAM enable operations to each DDS
+        for dds, ram_profile in self.ram_profile_map:
+            dds.set_cfr1(ram_enable=1,
+                         ram_destination=ram_profile.dest,
+                         osk_enable=ram_profile.osk_enable)
+
+        # Enable the RAM operations using the same I/O update pulse
+        for cpld in self.cplds:
+            cpld.io_update.pulse_mu(8)
+
+
+class AD9910RAM(JaxExperiment):
```

Comment:
I think this class should be a part of the example module? i.e. it is not something the user should directly use.